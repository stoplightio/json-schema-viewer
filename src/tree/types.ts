import type {
  ReferenceNode,
  RegularNode,
  SchemaNodeKind,
  SchemaTreeRefDereferenceFn,
} from '@stoplight/json-schema-tree';
import type { Optional } from '@stoplight/types';

import type { ViewMode } from '../types';

export type SchemaTreeRefInfo = {
  source: string | null;
  pointer: string | null;
};

export type SchemaTreeOptions = {
  expandedDepth: number;
  mergeAllOf: boolean;
  resolveRef: Optional<SchemaTreeRefDereferenceFn>;
  viewMode?: ViewMode;
};

export type ArrayNode = RegularNode & {
  primaryType: SchemaNodeKind.Array;
};

export type DictionaryNode = RegularNode & {
  primaryType: SchemaNodeKind.Object;
  fragment: {
    additionalProperties: Record<string, unknown>;
    [key: string]: unknown;
  };
};

export type PrimitiveArrayNode = ArrayNode & {
  children: [RegularNode & { simple: true }];
};

export type PrimitiveDictionaryNode = DictionaryNode & {
  children: [RegularNode & { simple: true }];
};

export type ComplexArrayNode = ArrayNode & {
  children: [RegularNode & { simple: false }];
};

export type ComplexDictionaryNode = DictionaryNode & {
  children: [RegularNode & { simple: false }];
};

export type BrokenRefArrayNode = ArrayNode & {
  children: [ReferenceNode & { error: string }];
};

export type FlattenableNode = PrimitiveArrayNode | PrimitiveDictionaryNode | ComplexArrayNode | BrokenRefArrayNode;
