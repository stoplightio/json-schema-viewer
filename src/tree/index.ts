export { getNodeMetadata } from './metadata';
export { SchemaTree, SchemaTreeState, SchemaTreeRefDereferenceFn } from './tree';
