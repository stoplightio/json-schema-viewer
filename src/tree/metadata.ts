import { TreeListNode } from '@stoplight/tree-list';
import { JsonPath } from '@stoplight/types';
import { JSONSchema, SchemaNode, SchemaTreeListNode } from '../types';

export interface ITreeNodeMetaSchema {
  path: JsonPath;
  schemaNode: SchemaNode;
  schema: JSONSchema;
}

export interface ITreeNodeMetaError {
  path: JsonPath;
  error: string;
}

export type TreeNodeMeta = ITreeNodeMetaSchema | ITreeNodeMetaError;

export const metadataStore = new WeakMap<SchemaTreeListNode, TreeNodeMeta>();

export const getNodeMetadata = (node: TreeListNode): TreeNodeMeta => {
  const metadata = metadataStore.get(node);
  if (metadata === void 0) {
    throw new Error('Missing metadata');
  }

  return metadata;
};

export const getSchemaNodeMetadata = (node: TreeListNode): ITreeNodeMetaSchema => {
  const metadata = getNodeMetadata(node);

  if (!('schema' in metadata)) {
    throw new TypeError('Schema node expected');
  }

  return metadata;
};
