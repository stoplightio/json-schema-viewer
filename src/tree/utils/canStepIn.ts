import { JSONSchema4 } from 'json-schema';

import { SchemaKind } from '../../types';
import { getCombiners } from '../../utils/getCombiners';
import { getPrimaryType } from '../../utils/getPrimaryType';

export const canStepIn = (fragment: JSONSchema4) => {
  if (getCombiners(fragment) !== void 0) {
    return true;
  }

  const type = getPrimaryType(fragment);
  return type === SchemaKind.Array || type === SchemaKind.Object;
};
