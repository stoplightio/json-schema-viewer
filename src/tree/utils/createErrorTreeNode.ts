import { TreeListParentNode } from '@stoplight/tree-list';

import { SchemaTreeListNode } from '../../types';
import { generateId } from '../../utils/generateId';
import { getSchemaNodeMetadata, metadataStore } from '../metadata';

export function createErrorTreeNode(parent: TreeListParentNode, error: string) {
  const { path } = getSchemaNodeMetadata(parent);

  const errorNode: SchemaTreeListNode = {
    id: generateId(),
    name: '',
    parent,
  };

  metadataStore.set(errorNode, {
    path,
    error,
  });

  return errorNode;
}
