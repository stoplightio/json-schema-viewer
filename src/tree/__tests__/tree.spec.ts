import { TreeListParentNode } from '@stoplight/tree-list';
import { JSONSchema4 } from 'json-schema';
import { ResolvingError } from '../../errors';
import { getNodeMetadata } from '../metadata';
import { SchemaTree, SchemaTreeState } from '../tree';
import { printTree } from './utils/printTree';

describe('SchemaTree', () => {
  describe('expanding', () => {
    describe('oneOf combiner', () => {
      let tree: SchemaTree;
      let schema: JSONSchema4;

      beforeEach(() => {
        schema = {
          type: 'object',
          properties: {
            user: {
              $ref: '#/properties/id',
            },
            id: {
              oneOf: [
                {
                  type: 'object',
                  required: ['foo'],
                  properties: {
                    foo: {
                      type: 'string',
                    },
                  },
                },
              ],
            },
          },
        };

        tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef: void 0,
        });

        tree.populate();
      });

      test('upon expanded $ref, should insert only oneOf combiner', () => {
        expect(tree.count).toEqual(3);

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);

        expect(tree.count).toEqual(4);
        expect(getNodeMetadata(tree.itemAt(2)!)).toHaveProperty(
          'schema',
          expect.objectContaining({
            oneOf: [
              {
                type: 'object',
                required: ['foo'],
                properties: {
                  foo: {
                    type: 'string',
                  },
                },
              },
            ],
          }),
        );
      });

      test('upon expanded $ref and expanded oneOf combiner, should insert object', () => {
        expect(tree.count).toEqual(3);

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);
        tree.unwrap(tree.itemAt(2) as TreeListParentNode);

        expect(tree.count).toEqual(5);
        expect(getNodeMetadata(tree.itemAt(3)!)).toHaveProperty(
          'schema',
          expect.objectContaining({
            type: 'object',
            required: ['foo'],
            properties: {
              foo: {
                type: 'string',
              },
            },
          }),
        );
      });

      test('upon expanded id property, should insert object', () => {
        expect(tree.count).toEqual(3);

        tree.unwrap(tree.itemAt(2) as TreeListParentNode);

        expect(tree.count).toEqual(4);
        expect(getNodeMetadata(tree.itemAt(3)!)).toHaveProperty(
          'schema',
          expect.objectContaining({
            type: 'object',
            required: ['foo'],
            properties: {
              foo: {
                type: 'string',
              },
            },
          }),
        );
      });
    });

    describe('array with $reffed items', () => {
      let tree: SchemaTree;
      let schema: JSONSchema4;

      beforeEach(() => {
        schema = {
          type: 'object',
          properties: {
            user: {
              type: 'array',
              items: {
                $ref: '#/properties/id',
              },
            },
            id: {
              type: 'object',
              required: ['foo'],
              properties: {
                foo: {
                  type: 'string',
                },
              },
            },
          },
        };

        tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef: void 0,
        });

        tree.populate();
      });

      test('upon expanded array, should insert object', () => {
        expect(tree.count).toEqual(3);

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);

        expect(tree.count).toEqual(4);
        expect(getNodeMetadata(tree.itemAt(2)!)).toHaveProperty(
          'schema',
          expect.objectContaining({
            type: 'object',
            required: ['foo'],
            properties: {
              foo: {
                type: 'string',
              },
            },
          }),
        );
      });
    });

    describe('empty $ref', () => {
      let schema: JSONSchema4;

      beforeEach(() => {
        schema = {
          type: 'object',
          properties: {
            id: {
              $ref: '',
            },
          },
        };
      });

      test('given no custom resolver, should generate an error', () => {
        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef: void 0,
        });

        tree.populate();

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(2) as TreeListParentNode)).toHaveProperty('error', 'The pointer is empty');
      });

      test('given a custom resolver, should attempt to resolve the reference', () => {
        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef() {
            throw new ResolvingError('Seems like you do not want this to be empty.');
          },
        });

        tree.populate();

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(2) as TreeListParentNode)).toHaveProperty(
          'error',
          'Seems like you do not want this to be empty.',
        );
      });
    });

    describe('external $refs', () => {
      let schema: JSONSchema4;

      beforeEach(() => {
        schema = {
          type: 'object',
          properties: {
            user: {
              type: 'array',
              items: {
                $ref: '../test#',
              },
            },
            id: {
              $ref: '../foo#id',
            },
          },
        };
      });

      test('given no custom resolver, should generate an error', () => {
        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef: void 0,
        });

        tree.populate();

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(2) as TreeListParentNode)).toHaveProperty(
          'error',
          'Cannot dereference external references',
        );

        tree.unwrap(tree.itemAt(3) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(4) as TreeListParentNode)).toHaveProperty(
          'error',
          'Cannot dereference external references',
        );
      });

      test('given a custom resolver, should attempt to resolve the reference', () => {
        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: 0,
          mergeAllOf: false,
          resolveRef({ source, pointer }) {
            if (source === '../test') {
              throw new ResolvingError(`Could not read "${source}"`);
            }

            throw new ResolvingError(`Pointer "${pointer}" is missing`);
          },
        });

        tree.populate();

        tree.unwrap(tree.itemAt(1) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(2) as TreeListParentNode)).toHaveProperty(
          'error',
          'Could not read "../test"',
        );

        tree.unwrap(tree.itemAt(3) as TreeListParentNode);
        expect(getNodeMetadata(tree.itemAt(4) as TreeListParentNode)).toHaveProperty(
          'error',
          'Pointer "#id" is missing',
        );
      });
    });

    describe('$refs in allOf', () => {
      test('given $ref in allOf', () => {
        const schema: JSONSchema4 = {
          type: 'object',
          properties: {
            foo: {
              type: 'object',
              properties: {
                name: {
                  type: 'string',
                },
                id: {
                  type: 'number',
                },
              },
            },
            bar: {
              allOf: [
                {
                  $ref: '#/properties/foo',
                },
                {
                  type: 'object',
                  properties: {
                    address: {
                      type: 'string',
                    },
                  },
                },
              ],
            },
          },
        };

        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: Infinity,
          mergeAllOf: true,
          resolveRef: void 0,
        });

        tree.populate();
        expect(printTree(tree)).toMatchInlineSnapshot(`
          "└─ #
             ├─ type: object
             └─ children
                ├─ 0
                │  └─ #/properties/foo
                │     ├─ type: object
                │     └─ children
                │        ├─ 0
                │        │  └─ #/properties/foo/properties/name
                │        │     └─ type: string
                │        └─ 1
                │           └─ #/properties/foo/properties/id
                │              └─ type: number
                └─ 1
                   └─ #/properties/bar
                      ├─ type: object
                      └─ children
                         ├─ 0
                         │  └─ #/properties/bar/properties/name
                         │     └─ type: string
                         ├─ 1
                         │  └─ #/properties/bar/properties/id
                         │     └─ type: number
                         └─ 2
                            └─ #/properties/bar/properties/address
                               └─ type: string
          "
        `);
      });

      test('given $ref in allOf pointing at another allOf, should keep merging', () => {
        const schema: JSONSchema4 = {
          type: 'object',
          properties: {
            baz: {
              type: 'object',
            },
            foo: {
              allOf: [
                {
                  type: 'object',
                  properties: {
                    name: {
                      type: 'string',
                    },
                    id: {
                      type: 'number',
                    },
                  },
                },
                {
                  $ref: '#/properties/baz',
                },
              ],
            },
            bar: {
              allOf: [
                {
                  $ref: '#/properties/foo',
                },
                {
                  type: 'object',
                  properties: {
                    address: {
                      type: 'string',
                    },
                  },
                },
              ],
            },
          },
        };

        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: Infinity,
          mergeAllOf: true,
          resolveRef: void 0,
        });

        tree.populate();
        expect(printTree(tree)).toMatchInlineSnapshot(`
          "└─ #
             ├─ type: object
             └─ children
                ├─ 0
                │  └─ #/properties/baz
                │     └─ type: object
                ├─ 1
                │  └─ #/properties/foo
                │     ├─ type: object
                │     └─ children
                │        ├─ 0
                │        │  └─ #/properties/foo/properties/name
                │        │     └─ type: string
                │        └─ 1
                │           └─ #/properties/foo/properties/id
                │              └─ type: number
                └─ 2
                   └─ #/properties/bar
                      ├─ type: object
                      └─ children
                         ├─ 0
                         │  └─ #/properties/bar/properties/name
                         │     └─ type: string
                         ├─ 1
                         │  └─ #/properties/bar/properties/id
                         │     └─ type: number
                         └─ 2
                            └─ #/properties/bar/properties/address
                               └─ type: string
          "
        `);
      });

      test('given direct circular reference pointing at allOf, should bail out and display unmerged allOf', () => {
        const schema: JSONSchema4 = {
          type: 'object',
          properties: {
            foo: {
              allOf: [
                {
                  $ref: '#/properties/bar',
                },
              ],
            },
            bar: {
              allOf: [
                {
                  $ref: '#/properties/foo',
                },
                {
                  type: 'object',
                  properties: {
                    name: {
                      type: 'string',
                    },
                  },
                },
              ],
            },
          },
        };

        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: Infinity,
          mergeAllOf: true,
          resolveRef: void 0,
        });

        expect(tree.populate.bind(tree)).not.toThrow();
        expect(printTree(tree)).toMatchInlineSnapshot(`
          "└─ #
             ├─ type: object
             └─ children
                ├─ 0
                │  └─ #/properties/foo
                │     ├─ type: undefined
                │     └─ children
                │        └─ 0
                │           └─ #/properties/foo/allOf/0
                │              ├─ type: undefined
                │              └─ children
                └─ 1
                   └─ #/properties/bar
                      ├─ type: undefined
                      └─ children
                         ├─ 0
                         │  └─ #/properties/bar/allOf/0
                         │     ├─ type: undefined
                         │     └─ children
                         └─ 1
                            └─ #/properties/bar/allOf/1
                               ├─ type: object
                               └─ children
                                  └─ 0
                                     └─ #/properties/bar/allOf/1/properties/name
                                        └─ type: string
          "
        `);
      });

      test('given indirect circular reference pointing at allOf, should bail out and display unmerged allOf', () => {
        const schema: JSONSchema4 = {
          type: 'object',
          properties: {
            baz: {
              allOf: [
                {
                  $ref: '#/properties/bar',
                },
              ],
            },
            foo: {
              allOf: [
                {
                  $ref: '#/properties/baz',
                },
              ],
            },
            bar: {
              allOf: [
                {
                  $ref: '#/properties/foo',
                },
                {
                  type: 'object',
                  properties: {
                    name: {
                      type: 'string',
                    },
                  },
                },
              ],
            },
          },
        };

        const tree = new SchemaTree(schema, new SchemaTreeState(), {
          expandedDepth: Infinity,
          mergeAllOf: true,
          resolveRef: void 0,
        });

        expect(tree.populate.bind(tree)).not.toThrow();
        expect(printTree(tree)).toMatchInlineSnapshot(`
          "└─ #
             ├─ type: object
             └─ children
                ├─ 0
                │  └─ #/properties/baz
                │     ├─ type: undefined
                │     └─ children
                │        └─ 0
                │           └─ #/properties/baz/allOf/0
                │              ├─ type: undefined
                │              └─ children
                ├─ 1
                │  └─ #/properties/foo
                │     ├─ type: undefined
                │     └─ children
                │        └─ 0
                │           └─ #/properties/foo/allOf/0
                │              ├─ type: undefined
                │              └─ children
                └─ 2
                   └─ #/properties/bar
                      ├─ type: undefined
                      └─ children
                         ├─ 0
                         │  └─ #/properties/bar/allOf/0
                         │     ├─ type: undefined
                         │     └─ children
                         └─ 1
                            └─ #/properties/bar/allOf/1
                               ├─ type: object
                               └─ children
                                  └─ 0
                                     └─ #/properties/bar/allOf/1/properties/name
                                        └─ type: string
          "
        `);
      });
    });
  });

  describe('allOf failures', () => {
    test('given incompatible values, should bail out and display unmerged allOf', () => {
      const schema: JSONSchema4 = {
        allOf: [
          {
            type: 'string',
          },
          {
            type: 'number',
          },
          {
            type: 'object',
            properties: {
              name: {
                type: 'string',
              },
            },
          },
        ],
      };

      const tree = new SchemaTree(schema, new SchemaTreeState(), {
        expandedDepth: Infinity,
        mergeAllOf: true,
        resolveRef: void 0,
      });

      expect(tree.populate.bind(tree)).not.toThrow();
      expect(printTree(tree)).toMatchInlineSnapshot(`
        "└─ #
           ├─ type: undefined
           └─ children
              ├─ 0
              │  └─ #/allOf/0
              │     └─ type: string
              ├─ 1
              │  └─ #/allOf/1
              │     └─ type: number
              └─ 2
                 └─ #/allOf/2
                    ├─ type: object
                    └─ children
                       └─ 0
                          └─ #/allOf/2/properties/name
                             └─ type: string
        "
      `);
    });
  });

  describe('paths generation', () => {
    let schema: JSONSchema4;
    let tree: SchemaTree;

    beforeEach(() => {
      schema = {
        type: 'object',
        properties: {
          user: {
            type: 'object',
            properties: {
              name: {
                type: 'string',
              },
              address: {
                type: 'string',
              },
            },
          },
          permissions: {
            type: 'array',
            items: {
              $ref: '#/properties/user',
            },
          },
        },
      };

      tree = new SchemaTree(schema, new SchemaTreeState(), {
        expandedDepth: Infinity,
        mergeAllOf: false,
        resolveRef: void 0,
      });
    });

    test('for root', () => {
      tree.populate();

      expect(getNodeMetadata(tree.itemAt(0)!)).toHaveProperty('path', []);
    });

    test('for plain object member', () => {
      tree.populate();

      expect(getNodeMetadata(tree.itemAt(1)!)).toHaveProperty('path', ['properties', 'user']);
    });

    test('for deep object member', () => {
      tree.populate();

      expect(getNodeMetadata(tree.itemAt(2)!)).toHaveProperty('path', ['properties', 'user', 'properties', 'name']);
    });

    test('for array items', () => {
      tree.populate();

      expect(getNodeMetadata(tree.itemAt(tree.count - 1)!)).toHaveProperty('path', ['properties', 'permissions']);
    });

    test('for $reffed array items', () => {
      tree.populate();
      tree.unwrap(tree.itemAt(tree.count - 1) as TreeListParentNode);

      expect(getNodeMetadata(tree.itemAt(tree.count - 3)!)).toHaveProperty('path', [
        'properties',
        'permissions',
        'items',
      ]);

      expect(getNodeMetadata(tree.itemAt(tree.count - 2)!)).toHaveProperty('path', [
        'properties',
        'permissions',
        'items',
        'properties',
        'name',
      ]);

      expect(getNodeMetadata(tree.itemAt(tree.count - 1)!)).toHaveProperty('path', [
        'properties',
        'permissions',
        'items',
        'properties',
        'address',
      ]);
    });
  });
});
