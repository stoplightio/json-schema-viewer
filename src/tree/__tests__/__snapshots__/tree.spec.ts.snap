// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaTree expanding $refs in allOf given very complex model with circular references, should bail out and display unmerged allOf 1`] = `
"└─ #
   ├─ type: object
   └─ children
      ├─ 0
      │  └─ #/properties/foo
      │     ├─ type: object
      │     └─ children
      │        └─ 0
      │           └─ #/properties/foo/properties/user
      │              ├─ combiner: allOf
      │              └─ children
      │                 └─ 0
      │                    └─ #/properties/foo/properties/user/allOf/0
      │                       ├─ type: object
      │                       └─ children
      │                          ├─ 0
      │                          │  └─ #/properties/foo/properties/user/allOf/0/properties/names
      │                          │     ├─ type: array
      │                          │     ├─ subtype: $ref[#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name]
      │                          │     └─ children
      │                          └─ 1
      │                             └─ #/properties/foo/properties/user/allOf/0/properties/users
      │                                ├─ type: array
      │                                ├─ subtype: object
      │                                └─ children
      │                                   ├─ 0
      │                                   │  └─ #/properties/foo/properties/user/allOf/0/properties/users/items/properties/creation
      │                                   │     ├─ $ref: #/allOf/0/allOf/0/properties/foo
      │                                   │     └─ children
      │                                   ├─ 1
      │                                   │  └─ #/properties/foo/properties/user/allOf/0/properties/users/items/properties/foo
      │                                   │     ├─ $ref: #/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts
      │                                   │     └─ children
      │                                   └─ 2
      │                                      └─ #/properties/foo/properties/user/allOf/0/properties/users/items/properties/products
      │                                         ├─ $ref: #/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts
      │                                         └─ children
      └─ 1
         └─ #/properties/bar
            ├─ type: object
            └─ children
               └─ 0
                  └─ #/properties/bar/properties/foo
                     ├─ type: object
                     └─ children
                        └─ 0
                           └─ #/properties/bar/properties/foo/properties/user
                              ├─ combiner: allOf
                              └─ children
                                 └─ 0
                                    └─ #/properties/bar/properties/foo/properties/user/allOf/0
                                       ├─ type: object
                                       └─ children
                                          ├─ 0
                                          │  └─ #/properties/bar/properties/foo/properties/user/allOf/0/properties/names
                                          │     ├─ type: array
                                          │     ├─ subtype: $ref[#/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/name]
                                          │     └─ children
                                          └─ 1
                                             └─ #/properties/bar/properties/foo/properties/user/allOf/0/properties/users
                                                ├─ type: array
                                                ├─ subtype: object
                                                └─ children
                                                   ├─ 0
                                                   │  └─ #/properties/bar/properties/foo/properties/user/allOf/0/properties/users/items/properties/creation
                                                   │     ├─ $ref: #/allOf/0/allOf/0/properties/foo
                                                   │     └─ children
                                                   ├─ 1
                                                   │  └─ #/properties/bar/properties/foo/properties/user/allOf/0/properties/users/items/properties/foo
                                                   │     ├─ $ref: #/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts
                                                   │     └─ children
                                                   └─ 2
                                                      └─ #/properties/bar/properties/foo/properties/user/allOf/0/properties/users/items/properties/products
                                                         ├─ $ref: #/allOf/0/allOf/0/properties/foo/definitions/event/allOf/0/properties/contacts
                                                         └─ children
"
`;

exports[`SchemaTree tree correctness given anyOf combiner placed next to allOf given allOf merging disabled, should still merge 1`] = `
"├─ 0
│  └─ #
│     ├─ type: object
│     ├─ combiner: anyOf
│     └─ children
│        ├─ 0
│        │  └─ #/anyOf/0
│        │     ├─ type: object
│        │     ├─ combiner: allOf
│        │     └─ children
│        │        ├─ 0
│        │        │  └─ #/anyOf/0/allOf/0
│        │        │     ├─ type: object
│        │        │     ├─ required
│        │        │     │  └─ 0: type
│        │        │     └─ children
│        │        │        ├─ 0
│        │        │        │  └─ #/anyOf/0/allOf/0/properties/type
│        │        │        │     ├─ type: string
│        │        │        │     └─ enum
│        │        │        │        ├─ 0: admin
│        │        │        │        └─ 1: editor
│        │        │        └─ 1
│        │        │           └─ #/anyOf/0/allOf/0/properties/enabled
│        │        │              └─ type: boolean
│        │        └─ 1
│        │           └─ #/anyOf/0/allOf/1
│        │              ├─ type: object
│        │              └─ children
│        │                 ├─ 0
│        │                 │  └─ #/anyOf/0/allOf/1/properties/root
│        │                 │     └─ type: boolean
│        │                 ├─ 1
│        │                 │  └─ #/anyOf/0/allOf/1/properties/group
│        │                 │     └─ type: string
│        │                 └─ 2
│        │                    └─ #/anyOf/0/allOf/1/properties/expirationDate
│        │                       └─ type: string
│        └─ 1
│           └─ #/anyOf/1
│              ├─ type: object
│              ├─ combiner: allOf
│              └─ children
│                 ├─ 0
│                 │  └─ #/anyOf/1/allOf/0
│                 │     ├─ type: object
│                 │     ├─ required
│                 │     │  └─ 0: type
│                 │     └─ children
│                 │        ├─ 0
│                 │        │  └─ #/anyOf/1/allOf/0/properties/type
│                 │        │     ├─ type: string
│                 │        │     └─ enum
│                 │        │        ├─ 0: admin
│                 │        │        └─ 1: editor
│                 │        └─ 1
│                 │           └─ #/anyOf/1/allOf/0/properties/enabled
│                 │              └─ type: boolean
│                 └─ 1
│                    └─ #/anyOf/1/allOf/1
│                       ├─ type: object
│                       └─ children
│                          ├─ 0
│                          │  └─ #/anyOf/1/allOf/1/properties/supervisor
│                          │     └─ type: string
│                          └─ 1
│                             └─ #/anyOf/1/allOf/1/properties/key
│                                └─ type: string
└─ 1
   └─ #
      ├─ type: object
      ├─ combiner: allOf
      └─ children
         └─ 0
            └─ #/allOf/0
               ├─ type: object
               ├─ required
               │  └─ 0: type
               └─ children
                  ├─ 0
                  │  └─ #/allOf/0/properties/type
                  │     ├─ type: string
                  │     └─ enum
                  │        ├─ 0: admin
                  │        └─ 1: editor
                  └─ 1
                     └─ #/allOf/0/properties/enabled
                        └─ type: boolean
"
`;

exports[`SchemaTree tree correctness given anyOf combiner placed next to allOf given allOf merging enabled, should merge contents of allOf combiners 1`] = `
"└─ #
   ├─ type: object
   ├─ combiner: anyOf
   └─ children
      ├─ 0
      │  └─ #/anyOf/0
      │     ├─ type: object
      │     ├─ required
      │     │  └─ 0: type
      │     └─ children
      │        ├─ 0
      │        │  └─ #/anyOf/0/properties/type
      │        │     ├─ type: string
      │        │     └─ enum
      │        │        ├─ 0: admin
      │        │        └─ 1: editor
      │        ├─ 1
      │        │  └─ #/anyOf/0/properties/enabled
      │        │     └─ type: boolean
      │        ├─ 2
      │        │  └─ #/anyOf/0/properties/root
      │        │     └─ type: boolean
      │        ├─ 3
      │        │  └─ #/anyOf/0/properties/group
      │        │     └─ type: string
      │        └─ 4
      │           └─ #/anyOf/0/properties/expirationDate
      │              └─ type: string
      └─ 1
         └─ #/anyOf/1
            ├─ type: object
            ├─ required
            │  └─ 0: type
            └─ children
               ├─ 0
               │  └─ #/anyOf/1/properties/type
               │     ├─ type: string
               │     └─ enum
               │        ├─ 0: admin
               │        └─ 1: editor
               ├─ 1
               │  └─ #/anyOf/1/properties/enabled
               │     └─ type: boolean
               ├─ 2
               │  └─ #/anyOf/1/properties/supervisor
               │     └─ type: string
               └─ 3
                  └─ #/anyOf/1/properties/key
                     └─ type: string
"
`;

exports[`SchemaTree tree correctness given oneOf combiner placed next to allOf given allOf merging disabled, should still merge 1`] = `
"├─ 0
│  └─ #
│     ├─ type: object
│     ├─ combiner: oneOf
│     └─ children
│        ├─ 0
│        │  └─ #/oneOf/0
│        │     ├─ type: object
│        │     ├─ combiner: allOf
│        │     └─ children
│        │        ├─ 0
│        │        │  └─ #/oneOf/0/allOf/0
│        │        │     ├─ type: object
│        │        │     ├─ required
│        │        │     │  └─ 0: type
│        │        │     └─ children
│        │        │        ├─ 0
│        │        │        │  └─ #/oneOf/0/allOf/0/properties/type
│        │        │        │     ├─ type: string
│        │        │        │     └─ enum
│        │        │        │        ├─ 0: admin
│        │        │        │        └─ 1: editor
│        │        │        └─ 1
│        │        │           └─ #/oneOf/0/allOf/0/properties/enabled
│        │        │              └─ type: boolean
│        │        └─ 1
│        │           └─ #/oneOf/0/allOf/1
│        │              ├─ type: object
│        │              └─ children
│        │                 ├─ 0
│        │                 │  └─ #/oneOf/0/allOf/1/properties/root
│        │                 │     └─ type: boolean
│        │                 ├─ 1
│        │                 │  └─ #/oneOf/0/allOf/1/properties/group
│        │                 │     └─ type: string
│        │                 └─ 2
│        │                    └─ #/oneOf/0/allOf/1/properties/expirationDate
│        │                       └─ type: string
│        └─ 1
│           └─ #/oneOf/1
│              ├─ type: object
│              ├─ combiner: allOf
│              └─ children
│                 ├─ 0
│                 │  └─ #/oneOf/1/allOf/0
│                 │     ├─ type: object
│                 │     ├─ required
│                 │     │  └─ 0: type
│                 │     └─ children
│                 │        ├─ 0
│                 │        │  └─ #/oneOf/1/allOf/0/properties/type
│                 │        │     ├─ type: string
│                 │        │     └─ enum
│                 │        │        ├─ 0: admin
│                 │        │        └─ 1: editor
│                 │        └─ 1
│                 │           └─ #/oneOf/1/allOf/0/properties/enabled
│                 │              └─ type: boolean
│                 └─ 1
│                    └─ #/oneOf/1/allOf/1
│                       ├─ type: object
│                       └─ children
│                          ├─ 0
│                          │  └─ #/oneOf/1/allOf/1/properties/supervisor
│                          │     └─ type: string
│                          └─ 1
│                             └─ #/oneOf/1/allOf/1/properties/key
│                                └─ type: string
└─ 1
   └─ #
      ├─ type: object
      ├─ combiner: allOf
      └─ children
         └─ 0
            └─ #/allOf/0
               ├─ type: object
               ├─ required
               │  └─ 0: type
               └─ children
                  ├─ 0
                  │  └─ #/allOf/0/properties/type
                  │     ├─ type: string
                  │     └─ enum
                  │        ├─ 0: admin
                  │        └─ 1: editor
                  └─ 1
                     └─ #/allOf/0/properties/enabled
                        └─ type: boolean
"
`;

exports[`SchemaTree tree correctness given oneOf combiner placed next to allOf given allOf merging enabled, should merge contents of allOf combiners 1`] = `
"└─ #
   ├─ type: object
   ├─ combiner: oneOf
   └─ children
      ├─ 0
      │  └─ #/oneOf/0
      │     ├─ type: object
      │     ├─ required
      │     │  └─ 0: type
      │     └─ children
      │        ├─ 0
      │        │  └─ #/oneOf/0/properties/type
      │        │     ├─ type: string
      │        │     └─ enum
      │        │        ├─ 0: admin
      │        │        └─ 1: editor
      │        ├─ 1
      │        │  └─ #/oneOf/0/properties/enabled
      │        │     └─ type: boolean
      │        ├─ 2
      │        │  └─ #/oneOf/0/properties/root
      │        │     └─ type: boolean
      │        ├─ 3
      │        │  └─ #/oneOf/0/properties/group
      │        │     └─ type: string
      │        └─ 4
      │           └─ #/oneOf/0/properties/expirationDate
      │              └─ type: string
      └─ 1
         └─ #/oneOf/1
            ├─ type: object
            ├─ required
            │  └─ 0: type
            └─ children
               ├─ 0
               │  └─ #/oneOf/1/properties/type
               │     ├─ type: string
               │     └─ enum
               │        ├─ 0: admin
               │        └─ 1: editor
               ├─ 1
               │  └─ #/oneOf/1/properties/enabled
               │     └─ type: boolean
               ├─ 2
               │  └─ #/oneOf/1/properties/supervisor
               │     └─ type: string
               └─ 3
                  └─ #/oneOf/1/properties/key
                     └─ type: string
"
`;

exports[`SchemaTree tree correctness should match arrays/of-allofs.json 1`] = `
"└─ #
   ├─ type: object
   └─ children
      └─ 0
         └─ #/properties/array-all-objects
            ├─ type: array
            ├─ subtype: object
            └─ children
               ├─ 0
               │  └─ #/properties/array-all-objects/items/properties/foo
               │     └─ type: string
               └─ 1
                  └─ #/properties/array-all-objects/items/properties/bar
                     └─ type: string
"
`;
